Makefile

```diff
@@ -16,86 +17,114 @@ else
+run: generate fmt vet manifests secret ## Run a controller from your host.
 	go run -mod=vendor ./main.go -api-secret-path=$(PWD)/.secret
 
-# Install CRDs into a cluster
-install: manifests
-	kustomize build config/crd | kubectl apply -f -
-
-# Uninstall CRDs from a cluster
-uninstall: manifests
-	kustomize build config/crd | kubectl delete -f -
+docker-build: ## Build the docker image with the manager.
+	docker build . -t ${IMG}
 
-# Deploy controller in the configured Kubernetes cluster in ~/.kube/config
-deploy: manifests
-	cd config/manager && kustomize edit set image controller=${IMG}
-	kustomize build config/default | kubectl apply -f -
+docker-push: ## Push the docker image with the manager.
+	docker push ${IMG}
 
-# Generate manifests e.g. CRD, RBAC etc.
-manifests: controller-gen
-	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+##@ Deployment
 
-# Run go fmt against code
-fmt:
-	go fmt ./...
+install: manifests kustomize ## Install CRDs into the K8s cluster specified in ~/.kube/config.
+	$(KUSTOMIZE) build config/crd | kubectl apply -f -
 
-# Run go vet against code
-vet:
-	go vet ./...
+uninstall: manifests kustomize ## Uninstall CRDs from the K8s cluster specified in ~/.kube/config.
+	$(KUSTOMIZE) build config/crd | kubectl delete -f -
 
-# Generate code
-generate: controller-gen mockgen
-	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
-	go generate ./...
+deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
+	cd config/manager && kustomize edit set image controller=${IMG}
+	$(KUSTOMIZE) build config/default | kubectl apply -f -
 
-# Build the docker image
-docker-build:
-	docker build . -t ${IMG}
+undeploy: ## Undeploy controller from the K8s cluster specified in ~/.kube/config.
+	$(KUSTOMIZE) build config/default | kubectl delete -f -
 
-# Push the docker image
-docker-push:
-	docker push ${IMG}
 
-# find or download controller-gen
-# download controller-gen if necessary
-controller-gen:
-ifeq (, $(shell which controller-gen))
-	@{ \
+KUBEBUILDER = $(shell pwd)/bin/kubebuilder
+kubebuilder: ## Download kubebuilder locally if necessary.
+	@[ -f $(KUBEBUILDER) ] || { \
 	set -e ;\
-	CONTROLLER_GEN_TMP_DIR=$$(mktemp -d) ;\
-	cd $$CONTROLLER_GEN_TMP_DIR ;\
+	TMP_DIR=$$(mktemp -d) ;\
+	cd $$TMP_DIR ;\
 	go mod init tmp ;\
-	go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.5 ;\
-	rm -rf $$CONTROLLER_GEN_TMP_DIR ;\
+	GOBIN=$(PROJECT_DIR)/bin go get sigs.k8s.io/kubebuilder/v3/cmd@v$(KUBEBUILDER_VERSION) ;\
+	rm -rf $$TMP_DIR ;\
+	cd - && mv bin/cmd $(KUBEBUILDER) ;\
 	}
-CONTROLLER_GEN=$(GOBIN)/controller-gen
-else
-CONTROLLER_GEN=$(shell which controller-gen)
-endif
+
+CONTROLLER_GEN = $(shell pwd)/bin/controller-gen
+controller-gen: ## Download controller-gen locally if necessary.
+	$(call go-get-tool,$(CONTROLLER_GEN),sigs.k8s.io/controller-tools/cmd/controller-gen@v0.4.1)
```

Comment:
Could you please create a variable for this version number?

Reply:
Yes, that can be done. I didn't change it because this is what kb v3 generates and was thinking to make as little change as possible.

Reply:
agree, thanks